<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Direct inference of higher-order chromatin structure in individual cells from scRNA-seq and scATAC-seq with compartmap • compartmap</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Direct inference of higher-order chromatin structure in individual cells from scRNA-seq and scATAC-seq with compartmap">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">compartmap</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.13.05</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/compartmap.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/compartmap/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Direct inference of higher-order chromatin structure in individual cells from scRNA-seq and scATAC-seq with compartmap</h1>
                        <h4 data-toc-skip class="author">Ben
Johnson</h4>
            
            <h4 data-toc-skip class="date">19 May 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/compartmap/blob/dev/vignettes/compartmap.Rmd" class="external-link"><code>vignettes/compartmap.Rmd</code></a></small>
      <div class="d-none name"><code>compartmap.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start-with-example-data">Quick start with example data<a class="anchor" aria-label="anchor" href="#quick-start-with-example-data"></a>
</h2>
<div class="section level4">
<h4 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h4>
<p>The expected input for <em>compartmap</em> is a
<code>RangedSummarizedExperiment</code> object. These can be built using
the built-in function <code><a href="../reference/importBigWig.html">importBigWig()</a></code> if starting from
BigWigs (recommended for scRNA-seq) or from a feature level object like
a <code>SingleCellExperiment</code> with the <code>rowRanges</code> slot
populated with the GRanges for each feature (see below in the
examples).</p>
<p>As an example for the quick start, we will load an existing example
scRNA-seq data from K562. These data are derived from Johnson and Rhodes
et. al 2021 STORM-seq They have already been TF-IDF transformed with
<code><a href="../reference/transformTFIDF.html">transformTFIDF()</a></code>. See further down if starting from bigWigs
or a feature based object. See the full workflow below for more
details.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/huishenlab/compartmap" class="external-link">compartmap</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"k562_scrna_chr14"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="inferring-higher-order-chromatin-domains-at-the-group-and-single-cell-level">Inferring higher-order chromatin domains at the group and
single-cell level<a class="anchor" aria-label="anchor" href="#inferring-higher-order-chromatin-domains-at-the-group-and-single-cell-level"></a>
</h4>
<div class="section level5">
<h5 id="group-level-inference">Group level inference<a class="anchor" aria-label="anchor" href="#group-level-inference"></a>
</h5>
<p>Process chr14 of the example K562 scRNA-seq data and infer
higher-order chromatin at 1Mb resolution:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k562_compartments</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scCompartments.html">scCompartments</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_scrna_chr14</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"rna"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="single-cell-level-inference">Single-cell level inference<a class="anchor" aria-label="anchor" href="#single-cell-level-inference"></a>
</h5>
<p>To infer higher-order domains in single cells and quantifying sign
coherence with the bootstrapping procedure, you can run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sub-sample to 10 cells as an example</span></span>
<span><span class="va">k562_scrna_chr14.sub</span> <span class="op">&lt;-</span> <span class="va">k562_scrna_chr14</span><span class="op">[</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">k562_scrna_chr14</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="va">k562_compartments.boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scCompartments.html">scCompartments</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_scrna_chr14.sub</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num.bootstraps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"rna"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Flip the domain sign if the sign coherence is discordant in 80% of the bootstraps</span></span>
<span><span class="va">k562_compartments.boot.fix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixCompartments.html">fixCompartments</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_compartments.boot</span>,</span>
<span>  min.conf <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the first cell in the GRangesList object</span></span>
<span><span class="va">k562_compartments.boot.fix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## GRanges object with 89 ranges and 13 metadata columns:</span></span>
<span><span class="co">##                             seqnames              ranges strand |         pc</span></span>
<span><span class="co">##                                &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;</span></span>
<span><span class="co">##     chr14:19000000-19999999    chr14   19000000-19999999      * |  -1.857418</span></span>
<span><span class="co">##     chr14:20000000-20999999    chr14   20000000-20999999      * |  -1.191918</span></span>
<span><span class="co">##     chr14:21000000-21999999    chr14   21000000-21999999      * |  -0.909080</span></span>
<span><span class="co">##     chr14:22000000-22999999    chr14   22000000-22999999      * |  -0.609605</span></span>
<span><span class="co">##     chr14:23000000-23999999    chr14   23000000-23999999      * |  -0.509780</span></span>
<span><span class="co">##                         ...      ...                 ...    ... .        ...</span></span>
<span><span class="co">##   chr14:103000000-103999999    chr14 103000000-103999999      * | -0.0605680</span></span>
<span><span class="co">##   chr14:104000000-104999999    chr14 104000000-104999999      * |  0.0891695</span></span>
<span><span class="co">##   chr14:105000000-105999999    chr14 105000000-105999999      * |  0.2389070</span></span>
<span><span class="co">##   chr14:106000000-106999999    chr14 106000000-106999999      * |  0.3886444</span></span>
<span><span class="co">##   chr14:107000000-107349539    chr14 107000000-107349539      * |  0.5383819</span></span>
<span><span class="co">##                             compartments      score boot.open boot.closed</span></span>
<span><span class="co">##                              &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##     chr14:19000000-19999999       closed  -1.857418         1           9</span></span>
<span><span class="co">##     chr14:20000000-20999999       closed  -1.191918         0          10</span></span>
<span><span class="co">##     chr14:21000000-21999999       closed  -0.909080         0          10</span></span>
<span><span class="co">##     chr14:22000000-22999999       closed  -0.609605         0          10</span></span>
<span><span class="co">##     chr14:23000000-23999999       closed  -0.509780         0          10</span></span>
<span><span class="co">##                         ...          ...        ...       ...         ...</span></span>
<span><span class="co">##   chr14:103000000-103999999       closed -0.0605680         4           6</span></span>
<span><span class="co">##   chr14:104000000-104999999         open  0.0891695         9           1</span></span>
<span><span class="co">##   chr14:105000000-105999999         open  0.2389070        10           0</span></span>
<span><span class="co">##   chr14:106000000-106999999         open  0.3886444        10           0</span></span>
<span><span class="co">##   chr14:107000000-107349539         open  0.5383819        10           0</span></span>
<span><span class="co">##                              conf.est conf.est.lowerCI conf.est.upperCI</span></span>
<span><span class="co">##                             &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">##     chr14:19000000-19999999  0.788987         0.574032                1</span></span>
<span><span class="co">##     chr14:20000000-20999999  0.861234         0.679113                1</span></span>
<span><span class="co">##     chr14:21000000-21999999  0.861234         0.679113                1</span></span>
<span><span class="co">##     chr14:22000000-22999999  0.861234         0.679113                1</span></span>
<span><span class="co">##     chr14:23000000-23999999  0.861234         0.679113                1</span></span>
<span><span class="co">##                         ...       ...              ...              ...</span></span>
<span><span class="co">##   chr14:103000000-103999999  0.572247         0.311604         0.832889</span></span>
<span><span class="co">##   chr14:104000000-104999999  0.788987         0.574032         1.000000</span></span>
<span><span class="co">##   chr14:105000000-105999999  0.861234         0.679113         1.000000</span></span>
<span><span class="co">##   chr14:106000000-106999999  0.861234         0.679113         1.000000</span></span>
<span><span class="co">##   chr14:107000000-107349539  0.861234         0.679113         1.000000</span></span>
<span><span class="co">##                             flip.compartment flip.score flip.conf.est</span></span>
<span><span class="co">##                                    &lt;logical&gt;  &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">##     chr14:19000000-19999999            FALSE  -1.857418      0.788987</span></span>
<span><span class="co">##     chr14:20000000-20999999            FALSE  -1.191918      0.861234</span></span>
<span><span class="co">##     chr14:21000000-21999999            FALSE  -0.909080      0.861234</span></span>
<span><span class="co">##     chr14:22000000-22999999            FALSE  -0.609605      0.861234</span></span>
<span><span class="co">##     chr14:23000000-23999999            FALSE  -0.509780      0.861234</span></span>
<span><span class="co">##                         ...              ...        ...           ...</span></span>
<span><span class="co">##   chr14:103000000-103999999            FALSE -0.0605680      0.572247</span></span>
<span><span class="co">##   chr14:104000000-104999999            FALSE  0.0891695      0.788987</span></span>
<span><span class="co">##   chr14:105000000-105999999            FALSE  0.2389070      0.861234</span></span>
<span><span class="co">##   chr14:106000000-106999999            FALSE  0.3886444      0.861234</span></span>
<span><span class="co">##   chr14:107000000-107349539            FALSE  0.5383819      0.861234</span></span>
<span><span class="co">##                             flip.conf.est.upperCI flip.conf.est.lowerCI</span></span>
<span><span class="co">##                                         &lt;numeric&gt;             &lt;numeric&gt;</span></span>
<span><span class="co">##     chr14:19000000-19999999                     1              0.574032</span></span>
<span><span class="co">##     chr14:20000000-20999999                     1              0.679113</span></span>
<span><span class="co">##     chr14:21000000-21999999                     1              0.679113</span></span>
<span><span class="co">##     chr14:22000000-22999999                     1              0.679113</span></span>
<span><span class="co">##     chr14:23000000-23999999                     1              0.679113</span></span>
<span><span class="co">##                         ...                   ...                   ...</span></span>
<span><span class="co">##   chr14:103000000-103999999              0.832889              0.311604</span></span>
<span><span class="co">##   chr14:104000000-104999999              1.000000              0.574032</span></span>
<span><span class="co">##   chr14:105000000-105999999              1.000000              0.679113</span></span>
<span><span class="co">##   chr14:106000000-106999999              1.000000              0.679113</span></span>
<span><span class="co">##   chr14:107000000-107349539              1.000000              0.679113</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="visualization-of-inferred-chromatin-domains">Visualization of inferred chromatin domains<a class="anchor" aria-label="anchor" href="#visualization-of-inferred-chromatin-domains"></a>
</h4>
<p>Once the data have been processed at either the group or single-cell
level, one can visualize the results using the <code>plotAB</code>
function in <em>compartmap</em>. Notably, we can include the confidence
intervals and median, chromosome-wide confidence estimate derived from
the bootstrap procedure for sign coherence. At 50%, this suggests that
estimates are evenly split between open and closed states. This may be
due to data sparsity or heterogeneity in the data. One possible approach
to resolve this is to increase the number of bootstraps performed if
initially set low (e.g. 10). Alternatively, it may be a region that is
worth investigating for your data set.</p>
<p>Plot the “fixed” results in cell 1 from above with plotAB with the
confidence intervals and median confidence estimate:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAB.html">plotAB</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_compartments.boot.fix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  what <span class="op">=</span> <span class="st">"flip.score"</span>,</span>
<span>  with.ci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  median.conf <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="compartmap_files/figure-html/postprocessPlot-1.png" width="700"></p>
<p>It is known that sometimes, the domains may be inverted relative to
orthogonal data This is also true in Hi-C and scHi-C domain inference
One can invert all domains by setting <code>reverse = TRUE</code> in the
<code><a href="../reference/plotAB.html">plotAB()</a></code> call</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAB.html">plotAB</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_compartments.boot.fix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  what <span class="op">=</span> <span class="st">"flip.score"</span>,</span>
<span>  with.ci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  median.conf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reverse <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extraction-of-domain-inflections">Extraction of domain inflections<a class="anchor" aria-label="anchor" href="#extraction-of-domain-inflections"></a>
</h4>
<p>It is often of interest to extract the chromatin domain inflection
points as they transition from “open” to “closed” states to look for
nearby CTCF sites, etc. We can accomplish this task using the
<code>getDomainInflections</code> function in <em>compartmap.</em></p>
<p>Domain inflections can be used to look for nearby CTCF sites, etc.
Here we extract single-cell domain inflections:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k562_cell_1_inflections</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDomainInflections.html">getDomainInflections</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_compartments.boot.fix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  what <span class="op">=</span> <span class="st">"flip.score"</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  chrs <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show the inflection points</span></span>
<span><span class="va">k562_cell_1_inflections</span></span>
<span><span class="co">## GRanges object with 18 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]    chr14  29000000      *</span></span>
<span><span class="co">##    [2]    chr14  30999999      *</span></span>
<span><span class="co">##    [3]    chr14  35000000      *</span></span>
<span><span class="co">##    [4]    chr14  38999999      *</span></span>
<span><span class="co">##    [5]    chr14  40000000      *</span></span>
<span><span class="co">##    ...      ...       ...    ...</span></span>
<span><span class="co">##   [14]    chr14  84999999      *</span></span>
<span><span class="co">##   [15]    chr14  89000000      *</span></span>
<span><span class="co">##   [16]    chr14  94999999      *</span></span>
<span><span class="co">##   [17]    chr14 104000000      *</span></span>
<span><span class="co">##   [18]    chr14 107349539      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="importing-bigwigs-as-input-to-compartmap">Importing bigWigs as input to compartmap<a class="anchor" aria-label="anchor" href="#importing-bigwigs-as-input-to-compartmap"></a>
</h3>
<p>The currently recommended input files to <em>compartmap</em> for
scRNA-seq are single-cell bigWigs, though does work with a
feature/counts based object as demonstrated in the next section.
Single-cell bigWigs can be generated through several tools, such as <a href="https://deeptools.readthedocs.io/en/latest/" class="external-link">deeptools</a>. To
import bigWigs, we can use the <code><a href="../reference/importBigWig.html">importBigWig()</a></code> function in
<em>compartmap</em>. This will read in a bigWig file and optionally
summarize to an arbitrary bin size. The bin size used in the
<em>compartmap</em> manuscript was 1kb and is what we do here as
well.</p>
<p>We can import a list of bigWig files and merge them into a
<code>RangedSummarizedExperiment</code> object:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># list the example bigWigs</span></span>
<span><span class="va">bigwigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate the 1kb bins</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19.gr"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span>
<span><span class="va">kb_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/tileGenome.html" class="external-link">tileGenome</a></span><span class="op">(</span></span>
<span>  seqlengths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">hg19.gr</span><span class="op">)</span><span class="op">[</span><span class="st">"chr14"</span><span class="op">]</span>,</span>
<span>  tilewidth <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  cut.last.tile.in.chrom <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import</span></span>
<span><span class="va">bigwigs_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bigwigs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/importBigWig.html">importBigWig</a></span><span class="op">(</span><span class="va">x</span>,</span>
<span>    bins <span class="op">=</span> <span class="va">kb_bins</span>,</span>
<span>    summarize <span class="op">=</span> <span class="cn">TRUE</span>, genome <span class="op">=</span> <span class="st">"hg19"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine</span></span>
<span><span class="va">bigwig_rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">bigwigs_lst</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bigwig_rse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_1"</span>, <span class="st">"cell_2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="starting-with-a-feature-or-counts-based-object">Starting with a feature or counts-based object<a class="anchor" aria-label="anchor" href="#starting-with-a-feature-or-counts-based-object"></a>
</h3>
<p>In the cases where we do not have or can’t start with bigWigs
(e.g. scATAC), we can start with a feature-level or counts object
(e.g. <code>SingleCellExperiment</code>). The two things that must be
there are making sure <code>rowRanges</code> and <code>colnames</code>
are set for each feature and cell/sample. We will use the scATAC-seq
from K562 as an example of how the object should look, but will also
show one way to add <code>rowRanges</code> to a
<code>SingleCellExperiment</code>, which works the same way for a
<code>SummarizedExperiment</code> object.</p>
<p>Loading the scATAC-seq example data pre-processed using the
<code>csaw</code> package:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"k562_scatac_chr14"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show the overall object structure</span></span>
<span><span class="co"># NOTE that the colnames are also there and the assay name is 'counts'</span></span>
<span><span class="va">k562_scatac_chr14</span></span>
<span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 11404 279 </span></span>
<span><span class="co">## metadata(6): spacing width ... param final.ext</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(279): cell_1 cell_2 ... cell_287 cell_288</span></span>
<span><span class="co">## colData names(4): bam.files totals ext rlen</span></span></code></pre></div>
<p>Showing the <code>rowRanges</code> slot is a <code>GRanges</code> for
each feature</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">k562_scatac_chr14</span><span class="op">)</span></span>
<span><span class="co">## GRanges object with 11404 ranges and 0 metadata columns:</span></span>
<span><span class="co">##           seqnames              ranges strand</span></span>
<span><span class="co">##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##       [1]    chr14   19614351-19614500      *</span></span>
<span><span class="co">##       [2]    chr14   19614401-19614550      *</span></span>
<span><span class="co">##       [3]    chr14   19614451-19614600      *</span></span>
<span><span class="co">##       [4]    chr14   19614501-19614650      *</span></span>
<span><span class="co">##       [5]    chr14   19614551-19614700      *</span></span>
<span><span class="co">##       ...      ...                 ...    ...</span></span>
<span><span class="co">##   [11400]    chr14 106938701-106938850      *</span></span>
<span><span class="co">##   [11401]    chr14 106938751-106938900      *</span></span>
<span><span class="co">##   [11402]    chr14 106938801-106938950      *</span></span>
<span><span class="co">##   [11403]    chr14 107280901-107281050      *</span></span>
<span><span class="co">##   [11404]    chr14 107280951-107281100      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre></div>
<p>But if we don’t have <code>rowRanges</code> for something like a
<code>SingleCellExperiment</code> we are working with, we need to
generate them. Thus, we will show an example of how to add
<code>rowRanges</code> from a GTF file to a
<code>SingleCellExperiment</code>.</p>
<p>First we define a helper function for adding <code>rowRanges</code>
(modified from <a href="https://github.com/trichelab/velocessor/blob/master/R/import_plate_txis.R" class="external-link uri">https://github.com/trichelab/velocessor/blob/master/R/import_plate_txis.R</a>).
<strong>NOTE</strong>: you can modify the
<code>rtracklayer::import</code> to not subset to gene level if using
feature level information (e.g. a bed file of fragments for scATAC)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">getRowRanges</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gtf</span>, <span class="va">asys</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">gxs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">gtf</span><span class="op">)</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gxs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gxs</span><span class="op">$</span><span class="va">gene_id</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">gxs</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">asys</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Import the example HEK293T SingleCellExperiment from the SMART-seq3
paper:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ss3_umi_sce"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import the example GTF with the helper function</span></span>
<span><span class="va">gtf_rowranges</span> <span class="op">&lt;-</span> <span class="fu">getRowRanges</span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/grch38_91_hsapiens.gtf.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"compartmap"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="va">ss3_umi_sce</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add rowRanges to the SingleCellExperiment</span></span>
<span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">ss3_umi_sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gtf_rowranges</span></span></code></pre></div>
<p>We can now proceed to the next steps and run the <em>compartmap</em>
workflow.</p>
</div>
<div class="section level3">
<h3 id="running-the-compartmap-workflow">Running the compartmap workflow<a class="anchor" aria-label="anchor" href="#running-the-compartmap-workflow"></a>
</h3>
<p>Once we have data in a <code>RangedSummarizedExperiment</code> or
other type of <code>SummarizedExperiment</code> with the
<code>rowRanges</code> slot filled, we can proceed through the
<em>compartmap</em> workflow. We will use the same K562 scRNA-seq data
shown in the manuscript on chromosome 14 here as the example.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example K562 data imported using importBigWig</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"k562_scrna_raw"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># TF-IDF transform the signals</span></span>
<span><span class="va">k562_scrna_chr14_tfidf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transformTFIDF.html">transformTFIDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">k562_scrna_se_chr14</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add back the TF-IDF counts to the object in the counts slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">k562_scrna_se_chr14</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">k562_scrna_chr14_tfidf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute chromatin domains at the group level</span></span>
<span><span class="va">k562_scrna_chr14_raw_domains</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scCompartments.html">scCompartments</a></span><span class="op">(</span><span class="va">k562_scrna_se_chr14</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num.bootstraps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"rna"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For single-cells, run:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k562_scrna_chr14_raw_domains</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scCompartments.html">scCompartments</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_scrna_se_chr14</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num.bootstraps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"rna"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>‘Fix’ compartments with discordant sign coherence:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k562_scrna_chr14_raw_domains.fix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixCompartments.html">fixCompartments</a></span><span class="op">(</span><span class="va">k562_scrna_chr14_raw_domains</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot results</span></span>
<span><span class="fu"><a href="../reference/plotAB.html">plotAB</a></span><span class="op">(</span><span class="va">k562_scrna_chr14_raw_domains.fix</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  what <span class="op">=</span> <span class="st">"flip.score"</span>,</span>
<span>  with.ci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  median.conf <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="compartmap_files/figure-html/workflow2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="higher-order-chromatin-interaction-maps">Higher-order chromatin interaction maps<a class="anchor" aria-label="anchor" href="#higher-order-chromatin-interaction-maps"></a>
</h3>
<p>Another interesting aspect we can derive from scRNA and scATAC is the
higher-order interacting domains through denoising of the correlation
matrices using a Random Matrix Theory approach. This is often
represented with the “plaid-like” patterning shown in Hi-C and scHi-C
approaches where stronger correlations (e.g. greater intensity of red)
indicates interacting domains relative to lesser correlation. We can do
something similar here through iterative denoising using Random Matrix
Theory approach:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k562_scrna_chr14_rmt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDenoisedMatrix.html">getDenoisedCorMatrix</a></span><span class="op">(</span></span>
<span>  <span class="va">k562_scrna_chr14</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr14"</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"rna"</span>,</span>
<span>  iter <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Shrinking bins with the JSE.</span></span>
<span><span class="co">## Number of means fewer than 4. Using Bayes instead of JSE.</span></span>
<span><span class="co">## 108 bins created...</span></span>
<span><span class="co">## Calculating correlations...</span></span>
<span><span class="co">## Done...</span></span>
<span><span class="co">## Denoising the correlation matrix using RMT.</span></span>
<span><span class="co">## Iterative denoising. Iteration: 2</span></span>
<span></span>
<span><span class="co"># Plot the interaction map</span></span>
<span><span class="fu"><a href="../reference/plotCorMatrix.html">plotCorMatrix</a></span><span class="op">(</span><span class="va">k562_scrna_chr14_rmt</span>, uppertri <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="compartmap_files/figure-html/denoisermt-1.png" width="700"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Benjamin Johnson, James Eapen, Tim Triche, Hui Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
