<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ compartmap</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">compartmap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.05</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/compartmap.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/compartmap/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="compartment-mapping">Compartment mapping<a class="anchor" aria-label="anchor" href="#compartment-mapping"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="scCompartments.html">scCompartments()</a></code>

        </dt>
        <dd>Estimate A/B compartments from single-cell sequencing data</dd>
      </dl><dl><dt>

          <code><a href="getArrayABsignal.html">getArrayABsignal()</a></code>

        </dt>
        <dd>Estimate A/B compartments from methylation array data</dd>
      </dl></div><div class="section level2">
      <h2 id="compartment-helper">Compartment helper<a class="anchor" aria-label="anchor" href="#compartment-helper"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="filterCompartments.html">filterCompartments()</a></code>

        </dt>
        <dd>Filter compartments using confidence estimates and eigenvalue thresholds</dd>
      </dl><dl><dt>

          <code><a href="fixCompartments.html">fixCompartments()</a></code>

        </dt>
        <dd>Invert, or "fix", compartments that have a minimum confidence score (1-min.conf)</dd>
      </dl><dl><dt>

          <code><a href="extractOpenClosed.html">extractOpenClosed()</a></code>

        </dt>
        <dd>Get the open and closed compartment calls based on sign of singular values</dd>
      </dl><dl><dt>

          <code><a href="getABSignal.html">getABSignal()</a></code>

        </dt>
        <dd>Calculate Pearson correlations of smoothed eigenvectors</dd>
      </dl><dl><dt>

          <code><a href="getATACABsignal.html">getATACABsignal()</a></code> <code><a href="getATACABsignal.html">getRNAABsignal()</a></code>

        </dt>
        <dd>Estimate A/B compartments from ATAC-seq data</dd>
      </dl></div><div class="section level2">
      <h2 id="correlation-matrix">Correlation matrix<a class="anchor" aria-label="anchor" href="#correlation-matrix"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="estRMT.html">estRMT()</a></code>

        </dt>
        <dd>Denoising of Covariance matrix using Random Matrix Theory</dd>
      </dl><dl><dt>

          <code><a href="getCorMatrix.html">getCorMatrix()</a></code>

        </dt>
        <dd>Calculate Pearson correlations of a binned matrix</dd>
      </dl><dl><dt>

          <code><a href="getDenoisedMatrix.html">getDenoisedCorMatrix()</a></code>

        </dt>
        <dd>Wrapper to denoise a correlation matrix using a Random Matrix Theory approach</dd>
      </dl></div><div class="section level2">
      <h2 id="data-transformation">Data transformation<a class="anchor" aria-label="anchor" href="#data-transformation"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="transformTFIDF.html">transformTFIDF()</a></code>

        </dt>
        <dd>Transform/normalize compartment calls using TF-IDF</dd>
      </dl><dl><dt>

          <code><a href="hdf5TFIDF.html">hdf5TFIDF()</a></code>

        </dt>
        <dd>Transform/normalize compartment calls using TF-IDF on HDF5-backed objects</dd>
      </dl></div><div class="section level2">
      <h2 id="bootstrapping">Bootstrapping<a class="anchor" aria-label="anchor" href="#bootstrapping"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="precomputeBootstrapMeans.html">precomputeBootstrapMeans()</a></code>

        </dt>
        <dd>Pre-compute the global means for bootstrapping compartments</dd>
      </dl><dl><dt>

          <code><a href="bootstrapCompartments.html">bootstrapCompartments()</a></code>

        </dt>
        <dd>Non-parametric bootstrapping of compartments and summarization of bootstraps/compute confidence intervals</dd>
      </dl><dl><dt>

          <code><a href="summarizeBootstraps.html">summarizeBootstraps()</a></code>

        </dt>
        <dd>Summarize the bootstrap compartment estimates and compute Agresti-Coull confidence intervals</dd>
      </dl></div><div class="section level2">
      <h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="importBigWig.html">importBigWig()</a></code>

        </dt>
        <dd>Import and optionally summarize a bigwig at a given resolution</dd>
      </dl><dl><dt>

          <code><a href="filterOpenSea.html">filterOpenSea()</a></code> <code><a href="filterOpenSea.html">getOpenSeas()</a></code>

        </dt>
        <dd>Filter to open sea CpG loci</dd>
      </dl><dl><dt>

          <code><a href="preprocessArrays.html">preprocessArrays()</a></code>

        </dt>
        <dd>Preprocess arrays for compartment inference</dd>
      </dl></div><div class="section level2">
      <h2 id="bioconductor-helpers">Bioconductor helpers<a class="anchor" aria-label="anchor" href="#bioconductor-helpers"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="condenseRE.html">condenseRE()</a></code>

        </dt>
        <dd>Condense a RaggedExperiment to a list of SummarizedExperiments</dd>
      </dl><dl><dt>

          <code><a href="condenseSE.html">condenseSE()</a></code>

        </dt>
        <dd>Condense the output of condenseRE to reconstruct per-sample GRanges objects to plot</dd>
      </dl><dl><dt>

          <code><a href="getAssayNames.html">getAssayNames()</a></code>

        </dt>
        <dd>Get the assay names from a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="getSeqLengths.html">getSeqLengths()</a></code>

        </dt>
        <dd>Get the seqlengths of a chromosome from a given genome's GRanges</dd>
      </dl><dl><dt>

          <code><a href="getChrs.html">getChrs()</a></code>

        </dt>
        <dd>Get the chromosomes from an object</dd>
      </dl></div><div class="section level2">
      <h2 id="stat-functions">Stat functions<a class="anchor" aria-label="anchor" href="#stat-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="agrestiCoullCI.html">agrestiCoullCI()</a></code>

        </dt>
        <dd>Agresti-Coull confidence interval for a binomial proportion</dd>
      </dl><dl><dt>

          <code><a href="fexpit.html">fexpit()</a></code>

        </dt>
        <dd>Helper function: expanded expit</dd>
      </dl><dl><dt>

          <code><a href="fisherZ.html">fisherZ()</a></code>

        </dt>
        <dd>Fisher's Z transformation</dd>
      </dl><dl><dt>

          <code><a href="flogit.html">flogit()</a></code>

        </dt>
        <dd>Helper function: squeezed logit</dd>
      </dl><dl><dt>

          <code><a href="getBinMatrix.html">getBinMatrix()</a></code>

        </dt>
        <dd>Generate bins for A/B compartment estimation</dd>
      </dl><dl><dt>

          <code><a href="getDomainInflections.html">getDomainInflections()</a></code>

        </dt>
        <dd>A wrapper function to generate a GRanges object of chromatin domain inflection points</dd>
      </dl><dl><dt>

          <code><a href="getGlobalMeans.html">getGlobalMeans()</a></code>

        </dt>
        <dd>Get the global means of a matrix</dd>
      </dl><dl><dt>

          <code><a href="getMatrixBlocks.html">getMatrixBlocks()</a></code>

        </dt>
        <dd>Get chunked sets of row-wise or column-wise indices of a matrix</dd>
      </dl><dl><dt>

          <code><a href="getSVD.html">getSVD()</a></code>

        </dt>
        <dd>Compute the SVD of a matrix using irlba</dd>
      </dl><dl><dt>

          <code><a href="getShrinkageTargets.html">getShrinkageTargets()</a></code>

        </dt>
        <dd>Get the specified samples to shrink towards instead of the global mean</dd>
      </dl><dl><dt>

          <code><a href="ifisherZ.html">ifisherZ()</a></code>

        </dt>
        <dd>Fisher's Z transformation</dd>
      </dl><dl><dt>

          <code><a href="imputeKNN.html">imputeKNN()</a></code>

        </dt>
        <dd>Impute missing values/NAs with KNN</dd>
      </dl><dl><dt>

          <code><a href="meanSmoother.html">meanSmoother()</a></code>

        </dt>
        <dd>Windowed mean smoother</dd>
      </dl><dl><dt>

          <code><a href="precomputeBootstrapMeans.html">precomputeBootstrapMeans()</a></code>

        </dt>
        <dd>Pre-compute the global means for bootstrapping compartments</dd>
      </dl><dl><dt>

          <code><a href="preprocessArrays.html">preprocessArrays()</a></code>

        </dt>
        <dd>Preprocess arrays for compartment inference</dd>
      </dl><dl><dt>

          <code><a href="removeEmptyBoots.html">removeEmptyBoots()</a></code>

        </dt>
        <dd>Remove bootstrap estimates that failed</dd>
      </dl><dl><dt>

          <code><a href="scCompartments.html">scCompartments()</a></code>

        </dt>
        <dd>Estimate A/B compartments from single-cell sequencing data</dd>
      </dl><dl><dt>

          <code><a href="shrinkBins.html">shrinkBins()</a></code>

        </dt>
        <dd>Employ an eBayes shrinkage approach for bin-level estimates for A/B inference</dd>
      </dl><dl><dt>

          <code><a href="sparseToDenseMatrix.html">sparseToDenseMatrix()</a></code>

        </dt>
        <dd>Convert a sparse matrix to a dense matrix in a block-wise fashion</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotAB.html">plotAB()</a></code>

        </dt>
        <dd>Plots A/B compartment estimates on a per chromosome basis</dd>
      </dl><dl><dt>

          <code><a href="plotCorMatrix.html">plotCorMatrix()</a></code>

        </dt>
        <dd>Plot a denoised correlation matrix</dd>
      </dl></div><div class="section level2">
      <h2 id="helpers">Helpers<a class="anchor" aria-label="anchor" href="#helpers"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="cleanAssayCols.html">cleanAssayCols()</a></code>

        </dt>
        <dd>Remove columns/cells/samples with NAs exceeding a threshold. See <code><a href="../reference/cleanAssay.html">cleanAssay()</a></code></dd>
      </dl><dl><dt>

          <code><a href="cleanAssayRows.html">cleanAssayRows()</a></code>

        </dt>
        <dd>Remove rows with NAs exceeding a threshold. See <code><a href="../reference/cleanAssay.html">cleanAssay()</a></code></dd>
      </dl><dl><dt>

          <code><a href="dot-n_approx.html">.n_approx()</a></code>

        </dt>
        <dd>n_tilde in AC</dd>
      </dl><dl><dt>

          <code><a href="dot-p_approx.html">.p_approx()</a></code>

        </dt>
        <dd>p_tilde in AC</dd>
      </dl><dl><dt>

          <code><a href="dot-z.html">.z()</a></code>

        </dt>
        <dd>Normal alpha/2 quantile</dd>
      </dl></div><div class="section level2">
      <h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="hg19.gr.html">hg19.gr</a></code>

        </dt>
        <dd>hg19 seqlengths as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="hg38.gr.html">hg38.gr</a></code>

        </dt>
        <dd>hg38 seqlengths as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="mm10.gr.html">mm10.gr</a></code>

        </dt>
        <dd>mm10 seqlengths as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="mm9.gr.html">mm9.gr</a></code>

        </dt>
        <dd>mm9 seqlengths as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="openSeas.hg19.html">openSeas.hg19</a></code>

        </dt>
        <dd>hg19 open sea CpG as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="openSeas.hg38.html">openSeas.hg38</a></code>

        </dt>
        <dd>hg38 open sea CpG as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="openSeas.mm10.html">openSeas.mm10</a></code>

        </dt>
        <dd>mm10 open sea CpG as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="openSeas.mm9.html">openSeas.mm9</a></code>

        </dt>
        <dd>mm9 open sea CpG as a GRanges object</dd>
      </dl><dl><dt>

          <code><a href="array.data.chr14.html">array_data_chr14</a></code>

        </dt>
        <dd>Example Illumina 450k methylation array data for compartmap</dd>
      </dl><dl><dt>

          <code><a href="k562_scatac_chr14.html">k562_scatac_chr14</a></code>

        </dt>
        <dd>Example scATAC-seq data for compartmap</dd>
      </dl><dl><dt>

          <code><a href="k562_scrna_chr14.html">k562_scrna_chr14</a></code>

        </dt>
        <dd>Example scRNA-seq data for compartmap</dd>
      </dl><dl><dt>

          <code><a href="k562_scrna_se_chr14.html">k562_scrna_raw</a></code>

        </dt>
        <dd>Example scRNA-seq data for compartmap</dd>
      </dl><dl><dt>

          <code><a href="ss3_umi_sce.html">ss3_umi_sce</a></code>

        </dt>
        <dd>Example SMART-seq3 scRNA-seq data for compartmap</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Benjamin Johnson, James Eapen, Tim Triche.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

