<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate A/B compartments from methylation array data — arrayCompartments • compartmap</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Estimate A/B compartments from methylation array data — arrayCompartments"><meta name="description" content="arrayCompartments returns estimated A/B compartments from methylation array data."><meta property="og:description" content="arrayCompartments returns estimated A/B compartments from methylation array data."><meta name="robots" content="noindex"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">compartmap</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.13.05</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/compartmap.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/compartmap/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate A/B compartments from methylation array data</h1>
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/compartmap/blob/dev/R/arrayCompartments.R" class="external-link"><code>R/arrayCompartments.R</code></a></small>
      <div class="d-none name"><code>arrayCompartments.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>arrayCompartments</code> returns estimated A/B compartments from methylation array data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">arrayCompartments</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">1000000</span>,</span>
<span>  chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  targets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num.bootstraps <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  preprocess <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  array.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hm450"</span>, <span class="st">"EPIC"</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19"</span>, <span class="st">"hg38"</span>, <span class="st">"mm9"</span>, <span class="st">"mm10"</span><span class="op">)</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boot.parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">bpparam</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>Input SummarizedExperiment object</p></dd>


<dt id="arg-res">res<a class="anchor" aria-label="anchor" href="#arg-res"></a></dt>
<dd><p>Compartment resolution in bp</p></dd>


<dt id="arg-chr">chr<a class="anchor" aria-label="anchor" href="#arg-chr"></a></dt>
<dd><p>What chromosome to work on (leave as NULL to run on all chromosomes)</p></dd>


<dt id="arg-group">group<a class="anchor" aria-label="anchor" href="#arg-group"></a></dt>
<dd><p>Whether to treat this as a group set of samples</p></dd>


<dt id="arg-targets">targets<a class="anchor" aria-label="anchor" href="#arg-targets"></a></dt>
<dd><p>Samples/cells to shrink towards</p></dd>


<dt id="arg-bootstrap">bootstrap<a class="anchor" aria-label="anchor" href="#arg-bootstrap"></a></dt>
<dd><p>Whether we should perform bootstrapping of inferred compartments</p></dd>


<dt id="arg-num-bootstraps">num.bootstraps<a class="anchor" aria-label="anchor" href="#arg-num-bootstraps"></a></dt>
<dd><p>How many bootstraps to run</p></dd>


<dt id="arg-preprocess">preprocess<a class="anchor" aria-label="anchor" href="#arg-preprocess"></a></dt>
<dd><p>Whether to preprocess the arrays prior to compartment inference</p></dd>


<dt id="arg-array-type">array.type<a class="anchor" aria-label="anchor" href="#arg-array-type"></a></dt>
<dd><p>What type of array is this ("hm450", "EPIC")</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>What genome to work on ("hg19", "hg38", "mm9", "mm10")</p></dd>


<dt id="arg-other">other<a class="anchor" aria-label="anchor" href="#arg-other"></a></dt>
<dd><p>Another arbitrary genome to compute compartments on</p></dd>


<dt id="arg-boot-parallel">boot.parallel<a class="anchor" aria-label="anchor" href="#arg-boot-parallel"></a></dt>
<dd><p>Whether to run the bootstrapping in parallel. See details.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>BiocParallelParam object to use for parallelization. See details.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A RaggedExperiment of inferred compartments</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>compartmap uses <code>BiocParallel</code> to parallelize operations in four
configurations. The default setting is to parallelize across columns but not
bootstraps using the thread count as reported by <code><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">BiocParallel::bpparam()</a></code>,
which is usually two cores fewer than the number of available cores.
Parallel bootstrapping is disabled by default to avoid nested parallelism
issues but can be done independent of column-wise parallelization.</p><div class="section">
<h3 id="available-configurations">Available configurations<a class="anchor" aria-label="anchor" href="#available-configurations"></a></h3>

<div class="section">
<h4 id="serial-bootstrapping">Serial bootstrapping<a class="anchor" aria-label="anchor" href="#serial-bootstrapping"></a></h4>

<ul><li><p>Serially with just one core:
<code>BPPARAM = BiocParallel::SerialParam()</code></p></li>
<li><p>Parallel across columns and serially across bootstraps:
<code>BPPARAM = BiocParallel::MulticoreParam(n)</code> where <code>n</code> is the number of
threads to use</p></li>
</ul><p>See <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">?BiocParallel::BiocParallelParam</a></code> for other parallel backends. Parallel
backends may also be passed to <code><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">BiocParallel::register()</a></code> to make them
available to <code>bpparam()</code>.</p>
</div>

<div class="section">
<h4 id="parallel-bootstrapping">Parallel bootstrapping<a class="anchor" aria-label="anchor" href="#parallel-bootstrapping"></a></h4>


<p>Set <code>boot.parallel = TRUE</code> for one the these configurations:</p><ul><li><p>Serially across columns and parallel across bootstraps: Set `BPPARAM =
list(SerialParam(), MulticoreParam(n))'</p></li>
<li><p>Parallel across both columns and bootstraps: Set <code>BPPARAM = list(MulticoreParam(outer), MulticoreParam(inner))</code> where <code>outer</code> is the
thread count for column-wise operations and <code>inner</code> the thread count for
bootstrapping. The required number of threads is given by</p></li>
</ul><p><code>( outer * inner ) + outer</code></p>
<p>which is more easily calculated as <code>outer * (inner + 1)</code>.</p>
<p>We recommend using an explicit list of two BiocParallelParam backends over
relying on <code>register()</code> and <code>bpparam()</code> for parallelizing across bootstraps.
With nested <code>bplapply</code> calls, the registered backend is used for both the
outer and inner parallel loops. On a system with 8 available threads if the
registered backend asks for 4 workers, it will try to use 20 threads in the
nested loops. Instead to use all 8 cores, set
<code>BPPARAM = list(MulticoreParam(2), MulticoreParam(3))</code>.</p>
</div>

<div class="section">
<h4 id="load-balancing">Load balancing<a class="anchor" aria-label="anchor" href="#load-balancing"></a></h4>


<p>Unless you have only 1 chromosome or are not bootstrapping/not bootstrapping
in parallel, you can use nested parallelism. If you are working on just 1
chromosome, put all cores into the inner bootstrapping backend. Conversely
with multiple chromosomes without bootstrapping, put all available workers in
the outer loop.</p>
<p>In general, use more 'outer' workers, which loop over chromosomes when <code>group = TRUE</code> and cells when <code>group = FALSE</code>, than 'inner' workers that loop over
bootstraps. Using 8 outer and 7 inner workers is faster than 7 outer and 8
inner.</p>
<p>When <code>group = FALSE</code>, use <code>MulticoreParam()</code> only on the outer workers. We
find that parallelizing at both column and bootstrap levels with the
single-cell inference is slower than only parallelizing at the column-level.</p>
<p>With <code>group = TRUE</code>, minimize the difference between the two worker counts:
with 64 total cores, doing 8 outer and 7 inner is faster than 16 outer and 3
inner.</p>
</div>


</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"minfi"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"array_data_chr14"</span>, package <span class="op">=</span> <span class="st">"compartmap"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">array_compartments</span> <span class="op">&lt;-</span> <span class="fu">arrayCompartments</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">array.data.chr14</span>,</span></span>
<span class="r-in"><span>    chr<span class="op">=</span><span class="st">"chr14"</span>,</span></span>
<span class="r-in"><span>    group<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    bootstrap<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    genome<span class="op">=</span><span class="st">"hg19"</span>,</span></span>
<span class="r-in"><span>    array.type<span class="op">=</span><span class="st">"hm450"</span>,</span></span>
<span class="r-in"><span>    BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting options('download.file.method.GEOquery'='auto')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Setting options('GEOquery.inmemory.gpl'=FALSE)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering to open sea CpG loci...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Dropping samples with &gt;80% NAs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Imputing missing data with kNN.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cluster size 3332 broken into 518 2814 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 518 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cluster size 2814 broken into 969 1845 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 969 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cluster size 1845 broken into 600 1245 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 600 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 1245 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 1845 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done cluster 2814 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> INFO [2026-02-25 20:05:58] Computing group level compartments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> INFO [2026-02-25 20:05:58] </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Benjamin Johnson, James Eapen, Tim Triche, Hui Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

